openhtf/io/frontend/src/app/station.ts

```diff
@@ -0,0 +1,138 @@
+
+
+/** Station view component. **/
+@Component({
+  selector: 'station',
+  templateUrl: 'templates/station.html',
+  styleUrls: ['styles/station.css'],
+  directives: [StationHeader, Prompt, TestHeader, PhaseListing, Logs],
+  providers: [StationService]
+})
+export class Station implements OnDestroy {
+  private framework: any = {};
+  private test: Test;
+  private stationState: any;
+  private lastUpdateTime: number;
+  private ip: string;
+  private port: string;
+  private countdown: Countdown;
+  private paused: boolean;
+  private pollingJob: number;
+
+  /**
+   * Create a Station view component.
+   */
+  constructor(private routeParams: RouteParams,
+              private stationService: StationService) {
+    this.paused = false;
+    this.stationState = {};
+    this.test = new Test();
+  }
+
+  /**
+   * Pause display updates for the given number of seconds.
+   */
+  pauseForCountdown(seconds: number) {
+    // TODO(jethier): Make the countdown timer look good. Possibly replace the
+    //                "current" button temporarily with a radial progress bar.
+    // TODO(jethier): Make this timeout configurable.
+    this.countdown = new Countdown(seconds);
+    this.paused = true;
+    this.countdown.onDone = () => { this.paused = false; };
+    this.countdown.reset();
+    this.countdown.start();
+  }
+
+  /**
+   * Set up the view component on initialization.
+   */
+  public ngOnInit(): any{
+    $('ul.tabs').tabs();
+    this.ip = this.routeParams.get('ip');
+    this.port = this.routeParams.get('port');
+    this.pollStationService();  // Avoid waiting for the first interval.
+    this.pollingJob = setInterval(() => this.pollStationService(), 250);
+  }
+
+  // TODO(jethier): Update this service to use pubsub instead of polling.
+  /**
+   * Make a request to the frontend server for this station's state.
+   */
+  pollStationService() {
+    // TODO(jethier): Fat arrow syntax below used to make __this unecessary.
+    //                Not sure when or why it broke. Investigation needed.
+    let __this = this;
+    this.stationService.getState(this.ip, this.port)
+                  .subscribe(
+                      state => __this.updateState(state),
+                      error => console.log('Error: ', error));
+  }
+
+  /**
+   * Use the given station state to update the view component.
+   */
+  updateState(state) {
+    // TODO(jethier): Detect when a new test is being started, display the
+    //                previous test, and pause for a countdown.
+    if (state && state.framework) {
+      this.framework = state.framework;
+      if (state.test) { 
+        this.test.update(state.test.record);
+
+        // Update the running phase.
+        let n = state.test.record.phases.length;
+        this.test.updatePhase(n, state.test.running_phase_record);
+
+        // Update pending phases using dummy phase records.
+        for (var i = 1; i < state.test.pending_phases.length; i++) {
```

Comment:
Easier to do it every time than to have a separate, _different_ update at the beginning. I could put it behind a flag to `updateState()`. But since the state comes as one big package you need that whole package anyway just to update the pending phases. Could implement run-once style logic or something. Definitely open to suggestions here.